contrib/endpoints/src/api_manager/config_manager.cc

```diff
@@ -0,0 +1,302 @@
+                                empty_configs);
+      return;
+    }
+
+    // TODO(jaebong) config_id should not be empty for the first version
+    // This part will be removed after the rollouts feature added
+    if (global_context_->config_id().empty()) {
+      std::string msg = "API config_id not specified in configuration files";
+      global_context_->env()->LogError(msg);
+      config_roollout_callback_(utils::Status(Code::ABORTED, msg),
+                                empty_configs);
+      return;
+    }
+
+    // Check config_id is invalid
+    if (!isValidConfigId(global_context_->config_id())) {
+      std::string msg = "Invalid config_id: " + global_context_->config_id();
+      global_context_->env()->LogError(msg);
+      config_roollout_callback_(utils::Status(Code::ABORTED, msg),
+                                empty_configs);
+      return;
+    }
+
+    std::function<void(utils::Status)> callback = [this](utils::Status status) {
+      on_fetch_auth_token(status);
+    };
+    // Fetch service account token
+    GlobalFetchServiceAccountToken(global_context_, callback);
+  } else if (status.code() == Code::INTERNAL) {
+    // Failed to fetch meta data
+    global_context_->env()->LogError("Metadata fetch failed.");
+    config_roollout_callback_(utils::Status(Code::ABORTED, status.message()),
+                              empty_configs);
+  } else {
+    // We should not get here
+    global_context_->env()->LogError("Unexpected status: " + status.ToString());
+    config_roollout_callback_(utils::Status(Code::ABORTED, status.message()),
+                              empty_configs);
+  }
+}
+
+void ConfigManager::on_fetch_auth_token(utils::Status status) {
+  if (status.ok()) {
+    if (global_context_->service_account_token()) {
+      // register auth token for servicemanagement services
+      global_context_->service_account_token()->SetAudience(
+          auth::ServiceAccountToken::JWT_TOKEN_FOR_SERVICEMANAGEMENT_SERVICES,
+          service_management_url_ + kServiceManagementServiceManager);
+    }
+
+    // Fetch configs from the Inception
+    std::shared_ptr<ConfigsFetchInfo> fetchInfo(new ConfigsFetchInfo);
+    // For now, config manager has only one config_id (100% rollout)
+    fetchInfo->rollouts = {{global_context_->config_id(), 0}};
+    fetch_configs(fetchInfo, 0);
+
+  } else if (status.code() == Code::INTERNAL) {
+    // Failed to fetch auth token.
+    global_context_->env()->LogError("Auth token fetch failed.");
+    config_roollout_callback_(utils::Status(Code::ABORTED, status.message()),
+                              empty_configs);
+  } else {
+    // We should not get here
+    global_context_->env()->LogError("Unexpected status: " + status.ToString());
+    config_roollout_callback_(utils::Status(Code::ABORTED, status.message()),
+                              empty_configs);
+  }
+}
+
+void ConfigManager::fetch_configs(std::shared_ptr<ConfigsFetchInfo> fetchInfo,
```

Comment:
can we move all fetch_configs logic into ConfigsFetchInfo class,   it will take a on_done callback,  ConfigManager will continue to execture on the on_done().